json.go

```diff
@@ -0,0 +1,556 @@
+		// If ExtJSON enabled and `handleExtJSON` returns a buffer, then this
+		// value was extended JSON and the value has been consumed.
+		if d.extJSONAllowed && outerTypeBytePos != topContainer {
+			buf, err := d.handleExtJSON(out, outerTypeBytePos)
+			if err != nil {
+				return nil, err
+			}
+			if buf != nil {
+				return buf, nil
+			}
+		}
+
+		// Not extended JSON, so now write the length placeholder
+		out = append(out, emptyLength...)
+		overwriteTypeByte(out, outerTypeBytePos, bsonDocument)
+
+		// Record position for the placeholder type byte
+		typeBytePos = len(out)
+		out = append(out, emptyType)
+
+		// Convert key as Cstring
+		out, err = d.convertCString(out)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, d.parseError(ch, "expecting key or end of object")
+	}
+
+	// Next non-WS char must be ':' for separator
+	err = d.readNameSeparator()
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert first value of object
+	out, err = d.convertValue(out, typeBytePos)
+	if err != nil {
+		return nil, err
+	}
+
+	// Loop, looking for separators or object terminator
+LOOP:
+	for {
+		ch, err = d.readAfterWS()
+		if err != nil {
+			return nil, newReadError(err)
+		}
+		switch ch {
+		case ',':
+			// Next non-WS character must be quote to start key
+			ch, err = d.readAfterWS()
+			if err != nil {
+				return nil, newReadError(err)
+			}
+			if ch != '"' {
+				return nil, d.parseError(ch, "expecting key")
+			}
+
+			// Record position for the placeholder type byte that we write
+			typeBytePos = len(out)
+			out = append(out, emptyType)
+
+			// Convert key as Cstring
+			out, err = d.convertCString(out)
+			if err != nil {
+				return nil, err
+			}
+
+			// Next non-WS char must be ':' for separator
+			err = d.readNameSeparator()
+			if err != nil {
+				return nil, err
+			}
+
+			// Convert next value
+			out, err = d.convertValue(out, typeBytePos)
+			if err != nil {
+				return nil, err
+			}
+		case '}':
+			break LOOP
+		default:
+			return nil, d.parseError(ch, "expecting value-separator or end of object")
+		}
+	}
+
+	// Write null terminator and calculate/update length
+	out = append(out, nullByte)
+	objectLength := len(out) - lengthPos
+	overwriteLength(out, lengthPos, objectLength)
+
+	return out, nil
+}
+
+// convertObject starts after the opening bracket of an array.
+func (d *Decoder) convertArray(out []byte) ([]byte, error) {
+	var ch byte
+	var err error
+
+	// Depth check
+	d.curDepth++
+	if d.curDepth > d.maxDepth {
+		return nil, errors.New("maximum depth exceeded")
+	}
+	defer func() { d.curDepth-- }()
+
+	// Note position of placeholder for length that we write
+	lengthPos := len(out)
+	out = append(out, emptyLength...)
+
+	ch, err = d.readAfterWS()
+	if err != nil {
+		return nil, newReadError(err)
+	}
+
+	// Case: empty array
+	if ch == ']' {
+		out = append(out, nullByte)
+		arrayLength := len(out) - lengthPos
+		overwriteLength(out, lengthPos, arrayLength)
+		return out, nil
+	}
```

Comment:
Given how this code is set up, I see why we special case the empty array here before entering the loop, but this makes the code more verbose than necessary for what seems to be a very small optimization. The `LOOP` could handle an empty array just like it handles non-empty (I believe the same applies to empty object in `convertObject` above). The body of the conditional I highlighted here is duplicated at the end of this method.

My suggestion will only save a few lines by getting rid of that duplication (pseudocode follows):
```
func (d) convertArray(...) {
    ...
    ch = d.readAfterWS()
    d.json.UnreadByte()

    if ch != ']' {
        // Not empty
        // Do all of the set-up for a non-empty array (lines 239 - 254)
    }

LOOP:
    for {
        ... // the body of the loop and the code that follows the loop will now handle the empty array just fine--no special casing needed.
    }

    // Write null terminator and calculate/update length
    out = append(out, nullByte)
    arrayLength := len(out) - lengthPos
    overwriteLength(out, lengthPos, arrayLength)  
}
```

Reply:
It's less about optimization and more about keeping the regular flow unintended and keeping the if block small.  (See [indent-error-flow](https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow) in the Google Code Review guidelines.)

While there's duplication reading the first entry outside the loop and the next entry inside the loop, we can't rely just on the `,` case of the switch -- we need duplication to address the first value without a comma.

That said, I've refactored the "convert an entry" into a method, which cuts down on the verbosity and improves consistency of the duplicated code.